timur/lib/client/jsx/components/model_map/model_report.jsx

```diff
@@ -137,38 +179,106 @@ const ModelReport = ({ model_name, updateCounts, counts, template, setAttribute
 
   const [ filterString, setFilterString ] = useState('');
 
+  const filterMatch = new RegExp(`^(?:(${ Object.keys(ATT_KEYS).join('|') }):)?(.*)$`)
+
   const matchesFilter = useCallback(attribute => {
     if (filterString == '') return true;
 
-    const { name, attribute_type, description } = attribute;
+    let tokens = filterString.split(/\s/).filter(_=>_).map(
+      token => token.match(filterMatch).slice(1)
+    );
 
-    if (filterString.split(/\s/).filter(_=>_).map(t => new RegExp(t,'i')).every(
-      token => (
-        name?.match(token) ||
-        attribute_type?.match(token) ||
-        description?.match(token)
-      )
-    )) return true;
+    return tokens.every( ([ column, token ]) => {
+      const tokenMatch = new RegExp(token,'i');
+      const values = (
+        column
+          ? [ attribute[ATT_KEYS[column]] ]
+          : Object.values(ATT_KEYS).map(
+            k => attribute[k]
+          )
+      );
 
-    return false;
+      return values.some( s => s?.match(tokenMatch));
+    });
   }, [ filterString ])
 
+  const [order, setOrder] = React.useState('asc');
+  const [orderBy, setOrderBy] = React.useState('type');
+
+  const sortByOrder = attributes => {
+    let srt;
+    if (orderBy === 'type') srt = sortAttributeList(attributes);
+    else {
+      srt = attributes.sort(
+        (a,b) => (a[ATT_KEYS[orderBy]]||'').localeCompare(b[ATT_KEYS[orderBy]]||'')
+      )
+    };
+    return (order === 'desc') ? srt.reverse() : srt;
+  }
+
+  const [ showDiff, setShowDiff ] = useState(false);
+  const [ diffProject, setDiffProject ] = useState(null);
+  const [ diffModel, setDiffModel ] = useState(null);
+  const [ diffTemplate, setDiffTemplate ] = useState(null);
+
+  const [anchor, setAnchor] = useState(null);
+
+  const getDiffModelTemplate = (project_name, model_name) => {
+    setDiffProject(project_name);
+    setDiffModel(model_name);
+    getDocuments(
+      {
+        project_name,
+        model_name,
+        record_names: [],
+        attribute_names: 'all'
+      },
+      fetch
+    ).then(({models}) => setDiffTemplate(models[model_name].template));
+  }
+  const attributes = Object.values({
+    ...template.attributes,
+    ...diffTemplate && diffTemplate.attributes
```

Comment:
hmm, does this expression work out?

assume `diffTemplate = null` => `...null && (null.attributes)` => `...null`.
I forget, but doesn't expanding a null result in an exception?

Reply:
Also, in javascript, `{}` is truthy, not falsy

Reply:
Seems to work as intended, ...falsy expands to nothing, i.e., `x = { ...null } => { }`.

Reply:
That's fascinating! Because I also thought `{...null}` would throw an exception, but it does seem to work in an assignment...

```
> {...null}
Uncaught SyntaxError: expected expression, got '...'
debugger eval code:1:1
> x = {...null}
Object {  }
```